### PR TITLE
[DevTestLabs] Adding export-artifacts commands on formula

### PR DESCRIPTION
**Pull request by vishrutshah**
fixes https://github.com/Azure/azure-cli/issues/2706

- [x] `export-artifacts` should export artifact array from the given formula
- [x] Exported artifacts should be usable for `az lab vm create` from formula

### CODE REVIEW

File: src/command_modules/azure-cli-lab/azure/cli/command_modules/lab/custom.py [hunk 1]
```diff
@@ -51,3 +51,24 @@ def list_vm(client, resource_group, lab_name, order_by=None, top=None,
...
                        expand=expand, filter=filters, top=top, order_by=order_by)
+
+
+def get_formula(client, resource_group, lab_name, name, export_artifacts=None):
+    """ Command to show formula from the Azure DevTest Lab """
+    formula = client.get_resource(resource_group_name=resource_group, lab_name=lab_name, name=name)
+    if export_artifacts:
+        return _export_artifacts(formula=formula)
+
+    return formula
```
Comment (troydai) [inline src/command_modules/azure-cli-lab/azure/cli/command_modules/lab/custom.py]:
```
return _export_artifacts(formula=forumla) if export_artifacts else formula
```

File: src/command_modules/azure-cli-lab/azure/cli/command_modules/lab/custom.py [hunk 2]
```diff
@@ -51,3 +51,24 @@ def list_vm(client, resource_group, lab_name, order_by=None, top=None,
...
+
+
+def _export_artifacts(formula):
```
Comment (troydai) [inline src/command_modules/azure-cli-lab/azure/cli/command_modules/lab/custom.py]: Make sure to document the size effects: because properties are deleted from each item of the `artifacts` list, which is __not__ a copy of `formula.formula_content.artifacts`, the original sources are changed as well. This is not a problem in current usage pattern because the command returns the formula and finishes immediately. But it will construct issues when this method is reused in other settings when the `formula.formula content.artifacts` is reused.

File: src/command_modules/azure-cli-lab/azure/cli/command_modules/lab/validators.py [hunk 3]
```diff
@@ -289,6 +290,33 @@ def _validate_other_parameters(namespace, formula=None):
             namespace.os_type = formula.os_type
 
 
+def _validate_artifacts(namespace):
```
Comment (troydai) [inline src/command_modules/azure-cli-lab/azure/cli/command_modules/lab/validators.py]: Maybe I'm late to the party. But the method name is misleading. It modifies the artifacts properties of namespace. The name `validate` indicate a read-only operations.

File: src/command_modules/azure-cli-lab/azure/cli/command_modules/lab/custom.py [hunk 4]
```diff
@@ -51,3 +51,24 @@ def list_vm(client, resource_group, lab_name, order_by=None, top=None,
 
     return client.list(resource_group, lab_name,
                        expand=expand, filter=filters, top=top, order_by=order_by)
+
+
+def get_formula(client, resource_group, lab_name, name, export_artifacts=None):
+    """ Command to show formula from the Azure DevTest Lab """
+    formula = client.get_resource(resource_group_name=resource_group, lab_name=lab_name, name=name)
```
Comment (tjprescott) [inline src/command_modules/azure-cli-lab/azure/cli/command_modules/lab/custom.py]: Alternatively, consider leaving the "show" operation as is and having "export-artifacts" as a convenience command. I think that would be more consistent with existing commands like this (i.e. KeyVault's get-default-policy).

Also, you could implement this by simply reflecting on the get operation and using a transformer method to massage the result (since that is really all _export_artifacts is doing). This would avoid the potential pitfall that @troydai mentions as well since it would be clear that this is an output transform only.

File: src/command_modules/azure-cli-lab/azure/cli/command_modules/lab/help.py [hunk 5]
```diff
@@ -87,7 +88,7 @@
                     az lab vm create --lab-name MyLab -g MyRG --name MyVM --image "Ubuntu Server 16.04 LTS" --image-type gallery --size Standard_DS1_v2 --ip-configuration public
                 - name: Create a Virtual Machine from a formula.
                   text: >
-                    az lab vm create --lab-name MyLab -g MyRG --name MyVM --formula MyFormula
+                    az lab vm create --lab-name MyLab -g MyRG --name MyVM --formula MyFormula --artifacts @/artifacts.json
```
Comment (tjprescott) [inline src/command_modules/azure-cli-lab/azure/cli/command_modules/lab/help.py]: Does this PR fix the fact that if you omit --artifacts it should use the formula's defaults?

File: src/command_modules/azure-cli-lab/azure/cli/command_modules/lab/validators.py [hunk 6]
```diff
@@ -162,7 +163,7 @@ def _validate_ip_configuration(namespace, lab_vnet=None):
             # Default to shared ip configuration based on os type only if inbound nat rules exist on the
             # shared configuration of the selected lab's virtual network
             if lab_vnet.subnet_overrides and lab_vnet.subnet_overrides[0].shared_public_ip_address_configuration and \
-                    lab_vnet.subnet_overrides[0].shared_public_ip_address_configuration.inbound_nat_rules:
+                    lab_vnet.subnet_overrides[0].shared_public_ip_address_configuration.allowed_ports:
```
Comment (tjprescott) [inline src/command_modules/azure-cli-lab/azure/cli/command_modules/lab/validators.py]: What's the reason for this change?

File: src/command_modules/azure-cli-lab/azure/cli/command_modules/lab/validators.py (hunk 3 above)
Comment (tjprescott) [inline src/command_modules/azure-cli-lab/azure/cli/command_modules/lab/validators.py]: You are a little late to the party 😁. I'm fine with this naming. It's called "validate" because it is used in the argument registration as a validator, so it is more natural when you type "validator=_validate_artifacts". It's pretty well known that validator is a terrible name since it is used for validation, transformation, etc. I would recommend sticking with the name.

File: src/command_modules/azure-cli-lab/azure/cli/command_modules/lab/help.py (hunk 5 above)
Comment (vishrutshah) [inline src/command_modules/azure-cli-lab/azure/cli/command_modules/lab/help.py]: Yup `_validate_artifacts` method would take care of using the defaults of `formula` while creating vm from `formula`

File: src/command_modules/azure-cli-lab/azure/cli/command_modules/lab/validators.py (hunk 6 above)
Comment (vishrutshah) [inline src/command_modules/azure-cli-lab/azure/cli/command_modules/lab/validators.py]: That's a bug :bug: Type of the `shared_public_ip_address_configuration` is of `SubnetSharedPublicIpAddressConfiguration` and not `SharedPublicIpAddressConfiguration`. That was discovered while manually testing out vm create. 

That is because i do not have scenario tests. I've also created task to be covered soon

**Comment by vishrutshah** [comment]
@tjprescott As per you suggestion, I've kept the normal `get` command on reflection and added `export-artifacts` with the transformer as separate convenience command using transformer. 

Yes i definitely need unit tests for all the validators, I'll adding all tests unit and scenario with the work item that I've created if that's fine for you. Let me know .

**Comment by tjprescott** [comment]
@vishrutshah it would be much better to add the relevant unit tests now rather than count on remembering to do them later. Scenario tests, that's fine, but the unit tests should be part of this PR, particularly testing that the logic to "complete the artifact ID" is covered and protected against regression. I've seen too many instances where "name or ID" logic breaks down when not protected by unit tests (from personal experience!)

**Comment by vishrutshah** [comment]
@tjprescott Agreed. On it :)
